backend/src/serverless/integrations/services/integrationProcessor.ts

```diff
@@ -289,49 +327,82 @@ export class IntegrationProcessor extends LoggingBase {
   }
 
   async process(req: NodeWorkerIntegrationProcessMessage) {
+    if (!req.runId) {
+      this.log.warn("No runId provided! Skipping because it's an old message.")
+      return
+    }
+
+    this.log.info({ runId: req.runId }, 'Detected integration run!')
+
+    const run = await this.integrationRunRepository.findById(req.runId)
+
+    const userContext = await getUserContext(run.tenantId)
+
+    // load integration from database
+    const integration = run.integrationId
+      ? await IntegrationRepository.findById(run.integrationId, userContext)
+      : await IntegrationRepository.findByPlatform(req.metadata.platform, userContext)
+
     const logger = createChildLogger('process', this.log, {
-      type: req.integrationType,
-      tenantId: req.tenantId,
-      integrationId: req.integrationId,
-      onboarding: req.onboarding,
-      microserviceId: req.microserviceId,
+      runId: req.runId,
+      type: integration.platform,
+      tenantId: integration.tenantId,
+      integrationId: run.integrationId,
+      onboarding: run.onboarding,
+      microserviceId: run.microserviceId,
     })
-    logger.info('Processing integration!')
 
-    const startOfProcessing = moment()
+    logger.info('Processing integration!')
 
-    const userContext = await getUserContext(req.tenantId)
     userContext.log = logger
 
-    // load integration from database
-    const integration = req.integrationId
-      ? await IntegrationRepository.findById(req.integrationId, userContext)
-      : await IntegrationRepository.findByPlatform(req.metadata.platform, userContext)
+    const existingRun = await this.integrationRunRepository.findLastProcessingRun(
+      run.integrationId,
+      run.microserviceId,
+      req.runId,
+    )
 
-    if (!req.onboarding) {
-      const processing = await this.redisCache.getValue(integration.id)
-      if (processing !== null) {
-        logger.info('Integration is already being processed!')
-        return
-      }
+    if (existingRun) {
+      logger.info('Integration is already being processed!')
+      await this.integrationRunRepository.markError(req.runId, {
+        message: 'Integration is already being processed!',
+        existingRunId: existingRun.id,
+      })
+      return
+    }
+
+    if (run.state === IntegrationRunState.PROCESSED) {
+      logger.warn('Integration is already processed!')
+      return
+    }
+
+    if (run.state === IntegrationRunState.PENDING) {
+      logger.info('Started processing integration!')
+    } else if (run.state === IntegrationRunState.DELAYED) {
+      logger.info('Continued processing delayed integration!')
+    } else if (run.state === IntegrationRunState.ERROR) {
+      logger.info('Restarted processing errored integration!')
+    } else if (run.state === IntegrationRunState.PROCESSING) {
+      throw new Error(`Invalid state '${run.state}' for integration run!`)
     }
```

Comment:
NIT: We're already checking processed state above, should this go in else block instead of else if?

Reply:
The above one is for `PROCESSED` state and this one is for `PROCESSING` state so we don't accidentally process the same run twice concurrently.